flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java

```diff
@@ -0,0 +1,211 @@
+
+		dataFileChannel = openFileChannel(dataFilePath);
+		indexFileChannel = openFileChannel(indexFilePath);
+	}
+
+	private FileChannel openFileChannel(Path path) throws IOException {
+		return FileChannel.open(path, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
+	}
+
+	/**
+	 * Persists the region index of the current data region and starts a new region to write.
+	 *
+	 * <p>Note: The caller is responsible for releasing the failed {@link PartitionedFile} if any exception
+	 * occurs.
+	 */
+	public void startNewRegion() throws IOException {
+		checkState(!isFinished, "File writer is already finished.");
+		checkState(dataFileChannel != null && indexFileChannel != null, "Must open the partitioned file first.");
+
+		writeRegionIndex();
+	}
+
+	private void writeRegionIndex() throws IOException {
+		if (Arrays.stream(subpartitionBytes).sum() > 0) {
+			for (int subpartition = 0; subpartition < numSubpartitions; ++subpartition) {
+				if (!indexBuffer.hasRemaining()) {
+					flushIndexBuffer();
+					indexBuffer.clear();
+					canCacheAllIndexData = false;
+				}
+
+				indexBuffer.putLong(totalBytesWritten);
+				indexBuffer.putInt(subpartitionBuffers[subpartition]);
+				totalBytesWritten += subpartitionBytes[subpartition];
+			}
+
+			++numRegions;
+			currentSubpartition = 0;
+			Arrays.fill(subpartitionBytes, 0);
+			Arrays.fill(subpartitionBuffers, 0);
+		}
+	}
+
+	private void flushIndexBuffer() throws IOException {
+		if (indexBuffer.position() > 0) {
+			indexBuffer.flip();
+			indexFileChannel.write(indexBuffer);
+		}
+	}
+
+	/**
+	 * Writes a {@link Buffer} of the given subpartition to the this {@link PartitionedFile}.
+	 *
+	 * <p>Note: The caller is responsible for recycling the target buffer and releasing the failed
+	 * {@link PartitionedFile} if any exception occurs.
+	 */
+	public void writeBuffer(Buffer target, int targetSubpartition) throws IOException {
+		checkArgument(targetSubpartition >= currentSubpartition, "Must write in subpartition index order.");
+		checkState(!isFinished, "File writer is already finished.");
+		checkState(dataFileChannel != null && indexFileChannel != null, "Must open the partitioned file first.");
+
+		currentSubpartition = Math.max(currentSubpartition, targetSubpartition);
+		long numBytes = BufferReaderWriterUtil.writeToByteChannel(dataFileChannel, target, header);
+
+		++subpartitionBuffers[targetSubpartition];
+		subpartitionBytes[targetSubpartition] += numBytes;
+	}
+
+	/**
+	 * Finishes writing which closes the file channel and returns the corresponding {@link PartitionedFile}.
+	 *
+	 * <p>Note: The caller is responsible for releasing the failed {@link PartitionedFile} if any exception
+	 * occurs.
+	 */
+	public PartitionedFile finish() throws IOException {
+		checkState(!isFinished, "File writer is already finished.");
+		checkState(dataFileChannel != null && indexFileChannel != null, "Must open the partitioned file first.");
```

Comment:
The `dataFileChannel` and `indexFileChannel` are created via data output driven. I am wondering that does it exist the case of empty output? Maybe the operator filters all the input data without any output, then finished finally.

If so, we should remove above check to avoid unnecessary failure.

Reply:
We always have an EndOfPartition event